rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/model/ParametricFileFacts.java

```diff
@@ -108,66 +134,228 @@ public void close(ISourceLocation loc) {
+            this.summaryBuilder = new ParametricSummaryBridge(exec, file, columns, contrib, lookupState, contrib::build);
         }
 
         public void reloadContributions() {
-            summary.reloadContributions();
+            summaryAnalyzer.reloadContributions();
+            summaryBuilder.reloadContributions();
         }
 
-        private void reportTypeCheckerMessages(List<Diagnostic> messages) {
-            typeCheckerMessages = messages;
-            sendDiagnostics();
+        private VersionedDiagnostics reportDiagnostics(AtomicReference<VersionedDiagnostics> box,
+                VersionedTree tree, List<Diagnostic> messages) {
+
+            var newer = new VersionedDiagnostics(tree, messages);
+            if (VersionedDiagnostics.setIfNewer(box, newer)) {
+                sendDiagnostics();
+            }
+            return newer;
         }
 
-        public void invalidate(boolean isClosing) {
+        private void invalidate(ParametricSummaryBridge summary, boolean isClosing) {
             summary.invalidate(isClosing);
         }
 
-        public void calculate() {
-            summary.calculateSummary();
-            summary.getMessages().thenAccept(this::reportTypeCheckerMessages);
+        public void invalidateAnalyzer(boolean isClosing) {
+            invalidate(summaryAnalyzer, isClosing);
+        }
+
+        public void invalidateBuilder(boolean isClosing) {
+            invalidate(summaryBuilder, isClosing);
+        }
+
+        /**
+         * @param version the version of the file for which summary calculation
+         * is currently requested
+         * @param latestVersion the last version of the file for which summary
+         * calculation was previously requested
+         * @param delay the duration after which the current request for summary
+         * calculation will be granted, unless another request is made in the
+         * meantime (in which case the current request is abandoned)
+         * @param calculation the actual summary calculation
+         * @return a future that provides optional diagnostics that result from
+         * the summary calculation
+         */
+        private CompletableFuture<Optional<VersionedDiagnostics>> calculate(
+                int version, AtomicInteger latestVersion, Duration delay,
+                Supplier<CompletableFuture<Optional<VersionedDiagnostics>>> calculation) {
+
+            latestVersion.set(version);
+            var delayed = CompletableFuture.delayedExecutor(delay.toMillis(), TimeUnit.MILLISECONDS, exec);
+            return CompletableFuture.supplyAsync(() -> {
+                // If no new call to `calculate` has been made after `delay` has
+                // passed (i.e., `lastVersion` hasn't changed in the meantime),
+                // then run the calculation. Else, abandon this calculation.
+                if (latestVersion.get() == version) {
```

Comment:
I like the simplicity of this debounce :+1:

Reply:
:)

Reply:
ah, it got even cleaner. good.